src/main/java/org/openjdk/jextract/impl/CDeclarationPrinter.java

```diff
@@ -0,0 +1,280 @@
+                    if (result.nameIncluded()) {
+                        return new TypeVisitorResult(true, result.typeStr());
+                    } else {
+                        return new TypeVisitorResult(false, result.typeStr() + "*");
+                    }
+                }
+                case UNSIGNED:
+                    return prefixedType("unsigned", t);
+                case SIGNED:
+                    return prefixedType("signed", t);
+                case VOLATILE:
+                    return prefixedType("volatile", t);
+                case COMPLEX:
+                    return prefixedType("complex", t);
+                case default:
+                    // defensive. If no name is present, we don't want to crash
+                    return new TypeVisitorResult(false,
+                        t.name().isPresent()? t.name().get() : t.toString());
+            }
+        }
+
+        @Override
+        public TypeVisitorResult visitFunction(Type.Function t, String context) {
+            String argsStr;
+            // Function type may optionally have parameter names.
+            // Include parameter names if available.
+            var optParameterNames = t.parameterNames();
+            if (optParameterNames.isPresent()) {
+                List<Type> argTypes = t.argumentTypes();
+                List<String> argNames = optParameterNames.get();
+                int numArgs = argTypes.size();
+                List<String> args = new ArrayList<>(numArgs);
+                for (int i = 0; i < numArgs; i++) {
+                    args.add(nameAndType(argTypes.get(i), argNames.get(i)));
+                }
+                argsStr = args.stream()
+                    .collect(Collectors.joining(",", "(", ")"));
+            } else {
+                argsStr = t.argumentTypes().stream()
+                    .map(a -> a.accept(this, "").typeStr())
+                    .collect(Collectors.joining(",", "(", ")"));
+            }
+            String res = t.returnType().accept(this, "").typeStr();
+            return new TypeVisitorResult(true, res + " (" + context + ")" + argsStr);
+        }
+
+        @Override
+        public TypeVisitorResult visitDeclared(Type.Declared t, String context) {
+            Declaration.Scoped scoped = t.tree();
+            return new TypeVisitorResult(false, typeTag(scoped) + " " + scoped.name());
+        }
+
+        @Override
+        public TypeVisitorResult visitArray(Type.Array t, String context) {
+            String brackets = String.format(" %s[%s]", context,
+                t.elementCount().isPresent() ? t.elementCount().getAsLong() : "");
+            return new TypeVisitorResult(true, t.elementType().accept(this, "").typeStr() + brackets);
+        }
+
+        @Override
+        public TypeVisitorResult visitType(Type t, String context) {
+            return new TypeVisitorResult(false, t.getClass().getName());
```

Comment:
Above in visitDelegated the default name seems to be `t.toString`. Maybe these could be factored out in some `defaultNameFor(Type type)` helper method for clarity.